qiskit_machine_learning/neural_networks/sampler_qnn.py

```diff
@@ -0,0 +1,343 @@
+
+    @property
+    def interpret(self) -> Callable[[int], int | tuple[int, ...]] | None:
+        """Returns interpret function to be used by the neural network. If it is not set in
+        the constructor or can not be implicitly derived, then ``None`` is returned."""
+        return self._interpret
+
+    def set_interpret(
+        self,
+        interpret: Callable[[int], int | tuple[int, ...]] | None = None,
+        output_shape: int | tuple[int, ...] | None = None,
+    ) -> None:
+        """Change 'interpret' and corresponding 'output_shape'.
+
+        Args:
+            interpret: A callable that maps the measured integer to another unsigned integer or
+                tuple of unsigned integers. See constructor for more details.
+            output_shape: The output shape of the custom interpretation, only used in the case
+                where an interpret function is provided. See constructor
+                for more details.
+        """
+
+        # derive target values to be used in computations
+        self._output_shape = self._compute_output_shape(interpret, output_shape)
+        self._interpret = interpret if interpret is not None else lambda x: x
+
+    def _compute_output_shape(
+        self,
+        interpret: Callable[[int], int | tuple[int, ...]] | None = None,
+        output_shape: int | tuple[int, ...] | None = None,
+    ) -> tuple[int, ...]:
+        """Validate and compute the output shape."""
+
+        # this definition is required by mypy
+        output_shape_: tuple[int, ...] = (-1,)
+
+        if interpret is not None:
+            if output_shape is None:
+                raise QiskitMachineLearningError(
+                    "No output shape given, but required in case of custom interpret!"
+                )
+            if isinstance(output_shape, Integral):
+                output_shape = int(output_shape)
+                output_shape_ = (output_shape,)
+            else:
+                output_shape_ = output_shape  # type: ignore
+        else:
+            if output_shape is not None:
+                # Warn user that output_shape parameter will be ignored
+                logger.warning(
+                    "No interpret function given, output_shape will be automatically "
+                    "determined as 2^num_qubits."
+                )
+            output_shape_ = (2**self._circuit.num_qubits,)
+
+        return output_shape_
+
+    def _preprocess(
+        self,
+        input_data: list[float] | np.ndarray | float,
+        weights: list[float] | np.ndarray | float,
+    ) -> tuple[np.ndarray, int]:
+        """
+        Pre-processing during forward pass of the network.
+        """
+
+        np_input_data: np.ndarray = np.asarray(input_data)
+        np_weights: np.ndarray = np.asarray(weights)
+        if len(np_input_data.shape) == 1:
+            np_input_data = np.expand_dims(np_input_data, 0)
+
+        num_samples = max(np_input_data.shape[0], 1)
```

Comment:
this max is there to deal with the case where `input_data=[]`, but `weights != []`, because `num_samples` is used later to broadcast the weights in the sampler call. I could also remove the max here and check for `num_samples==0` in the forward pass, but I thought this would be more compact.